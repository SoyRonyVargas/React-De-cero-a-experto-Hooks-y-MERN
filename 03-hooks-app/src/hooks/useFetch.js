import { useEffect, useState } from 'react'

const useFetch = (url) => {
    
    const [ state , setState ] = useState({
        data: null,
        loading: true,
        error: null
    })

    useEffect(() => {
        
        getData()

        return () => {

        }

    }, [url])

    const getData = async () => {

        setState({
            ...state,
            loading: true
        })

        const response = await fetch(url)

        const data = await response.json()

        setTimeout( () => {
            setState({
                ...state,
                data: data,
                loading: false
            })
        }, 0)

    }

    return state
}

export default useFetch
