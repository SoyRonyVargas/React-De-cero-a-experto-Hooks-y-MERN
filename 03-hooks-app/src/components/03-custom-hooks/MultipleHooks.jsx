import useCounter from '../../hooks/useCounter'
import useFetch from '../../hooks/useFetch'
import React from 'react'

const MultipleHooks = () => {

    const {
        counter,
        increment
    } = useCounter(1)

    const {
        data,
        loading
    } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`)

    const { author , quote } = !!data && data[0]

    return (
        <div className="">
            <h1>Breaking Bad Quotes</h1>
            <hr />
            {
                !loading ?
                <figure className="card p-3">
                <blockquote className="blockquote">
                    <p> {quote} </p>
                </blockquote>
                <figcaption className="blockquote-footer">
                    { author }
                </figcaption>
                </figure>
                :
                <div className="alert alert-info text-center">
                    Loading...
                </div>
            }
            <button onClick={increment} className="btn btn-primary">
                Next Quote
            </button>
        </div>
    )
}

export default MultipleHooks
