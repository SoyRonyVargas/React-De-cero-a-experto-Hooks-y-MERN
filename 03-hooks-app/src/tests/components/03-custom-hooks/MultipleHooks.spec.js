import MultipleHooks from '../../../components/03-custom-hooks/MultipleHooks'
import useCounter from '../../../hooks/useCounter'
import useFetch from '../../../hooks/useFetch'
jest.mock('../../../hooks/useCounter')
jest.mock('../../../hooks/useFetch')
import { shallow } from 'enzyme'
import React from 'react'

describe('Probando el componente multiple custom hooks', () => { 
    
    beforeEach( () => {
        
        useCounter.mockReturnValue({
            counter: 1,
            increment: () => {},
        });

    })
    
    test('deberia mostrarse correctamente el componente', () => { 
        

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })

        const wrapper = shallow(<MultipleHooks/>)

        expect(wrapper).toMatchSnapshot()

    })

    test('debe de mostrar la informaciÃ³n', () => { 

        useFetch.mockReturnValue({
            data: [
                {
                    quote: 'test',
                    author: 'Mario palacios'
                }
            ],
            loading: false,
            error: null
        })

        const wrapper = shallow(<MultipleHooks/>)

        expect(wrapper.find('.alert').exists()).toBe(false)
        
        console.log(wrapper.html());

    })
    
})