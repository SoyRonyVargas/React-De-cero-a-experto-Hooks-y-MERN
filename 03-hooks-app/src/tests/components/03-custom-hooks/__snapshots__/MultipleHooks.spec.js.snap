// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente multiple custom hooks deberia mostrarse correctamente el componente 1`] = `
<div
  className=""
  id="container"
>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next Quote
  </button>
</div>
`;
